### PR TITLE
[AutoMM] Parameter Efficient Fine-tuning SAM with LoRA for binary semantic segmentation tasks.

### PR DESCRIPTION
**Pull request by Harry-zzh**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/models/utils.py [hunk 1]
```diff
@@ -502,22 +506,27 @@ def inject_adaptation_to_linear_layer(
...
-
+    extra_trainable_params
+        Not to apply loRA to modules filtered by name, and these modules are not frozen during training (e.g. "mask_decoder").
+        If None, all the modules except for those applied loRA are frozen.
     Returns
     -------
     Model with injected LoRA modules.
     """
     for m_name, module in dict(model.named_modules()).items():
-        if not module_filter or any(re.match(filter_module, m_name) for filter_module in module_filter):
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/utils.py:511]: How about adding a if-continue here?

File: multimodal/src/autogluon/multimodal/problem_types.py [hunk 2]
```diff
@@ -20,6 +20,7 @@
     NUMERICAL,
     OBJECT_DETECTION,
     OPEN_VOCABULARY_OBJECT_DETECTION,
+    REAL_WORLD_SEM_SEG,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/problem_types.py:23]: Consider removing the real_world key.

File: core/src/autogluon/core/metrics/__init__.py [hunk 3]
```diff
@@ -553,10 +553,7 @@ def customized_log_loss(y_true, y_pred, eps=1e-15):
     for metric_dict in [BINARY_METRICS, MULTICLASS_METRICS]:
         _add_scorer_to_metric_dict(metric_dict=metric_dict, scorer=scorer)
 
-for scorer in [
-    roc_auc,
-    average_precision,
-]:
```
Comment (zhiqiangdon) [inline core/src/autogluon/core/metrics/__init__.py:559]: Maybe revert this change.

File: multimodal/src/autogluon/multimodal/data/process_semantic_seg_img.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+            image_transforms=self.val_transforms, size=self.size, normalization=self.normalization
+        )
+
+        self.gt_train_processor = construct_image_processor(
+            image_transforms=self.gt_train_transforms, size=self.size, normalization=None
+        )
+        self.gt_val_processor = construct_image_processor(
+            image_transforms=self.gt_val_transforms, size=self.size, normalization=None
+        )
+        self.augmentation = self.get_aug_transform()
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/process_semantic_seg_img.py:124]: augmentation operations should be in the `train_processor`.

File: multimodal/src/autogluon/multimodal/data/process_semantic_seg_img.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+        odict = self.__dict__.copy()  # get attribute dictionary
+        del odict["train_processor"]  # remove augmenter to support pickle
+        return odict
+
+    def __setstate__(self, state):
+        self.__dict__ = state
+        if "train_transform_types" in state:  # backward compatible
+            self.train_transforms = list(self.train_transform_types)
+        if "val_transform_types" in state:
+            self.val_transforms = list(self.val_transform_types)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/process_semantic_seg_img.py:282]: No need of these lines.

File: multimodal/src/autogluon/multimodal/data/utils.py [hunk 6]
```diff
@@ -583,7 +587,7 @@ def construct_image_processor(
     image_transforms = get_image_transform_funcs(transform_types=image_transforms, size=size)
     if not any([isinstance(trans, transforms.ToTensor) for trans in image_transforms]):
         image_transforms.append(transforms.ToTensor())
-    if not any([isinstance(trans, transforms.Normalize) for trans in image_transforms]):
+    if not any([isinstance(trans, transforms.Normalize) for trans in image_transforms]) and normalization != None:
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/utils.py:590]: When would be `normalization` None?

File: multimodal/src/autogluon/multimodal/optimization/losses.py [hunk 7]
```diff
@@ -340,3 +340,41 @@ def forward(self, input: torch.Tensor, target: torch.Tensor):
             loss = loss.sum()
 
         return loss
+
+
+class IoULoss(nn.Module):
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/optimization/losses.py:345]: Need some docstrings for the class.

File: multimodal/src/autogluon/multimodal/optimization/losses.py [hunk 8]
```diff
@@ -340,3 +340,41 @@ def forward(self, input: torch.Tensor, target: torch.Tensor):
...
+        input = torch.sigmoid(input)
+        inter = ((input * target) * weit).sum(dim=(2, 3))
+        union = ((input + target) * weit).sum(dim=(2, 3))
+        wiou = 1 - (inter + 1) / (union - inter + 1)
+        return (wbce + wiou).mean()
+
+
+class BBCEWithLogitLoss(nn.Module):
+    """
+    Balanced BCEWithLogitLoss
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/optimization/losses.py:362]: More detailed docstrings?

File: multimodal/src/autogluon/multimodal/optimization/semantic_segmentation_metrics/metric.py
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/optimization/semantic_segmentation_metrics/metric.py:1]: Do we need the `semantic_segmentation_metrics` folder? Maybe just one file like `semantic_seg_metrics.py`

File: multimodal/src/autogluon/multimodal/optimization/utils.py [hunk 9]
```diff
@@ -123,6 +133,13 @@ def get_loss_func(
         loss_func = nn.CrossEntropyLoss(ignore_index=0)
     elif problem_type in [OBJECT_DETECTION, OPEN_VOCABULARY_OBJECT_DETECTION]:
         return None
+    elif problem_type == SEMANTIC_SEGMENTATION:
+        if "iou_loss" in loss_func_name.lower():
+            loss_func = IoULoss()
+        elif "balanced_bce" in loss_func_name.lower():
+            loss_func = BBCEWithLogitLoss()
+        else:
+            loss_func = nn.BCEWithLogitsLoss()
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/optimization/utils.py:142]: Does pytorch have `BBCEWithLogitLoss`? Typo? Combine elif and else here?

File: multimodal/src/autogluon/multimodal/problem_types.py [hunk 10]
```diff
@@ -131,6 +132,19 @@ def fallback_label_type(self):
     ),
 )
 
+# Real-World Semantic Segmentation: image --> image
+PROBLEM_TYPES_REG.register(
+    SEMANTIC_SEGMENTATION,
+    ProblemTypeProperty(
+        name=SEMANTIC_SEGMENTATION,
+        support_zero_shot=True,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/problem_types.py:140]: Need to test zero-shot

File: multimodal/src/autogluon/multimodal/utils/data.py [hunk 11]
```diff
@@ -199,6 +202,10 @@ def create_data_processor(
             text_max_len=model_config.max_text_len,
             missing_value_strategy=config.data.document.missing_value_strategy,
         )
+    elif data_type == SEMANTIC_SEGMENTATION_IMG:
+        data_processor = SemanticSegImageProcessor(
+            model=model, model_config=model_config, norm_type=model_config.image_norm
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/data.py:207]: `model` already has the attribute config, which is the same as `model_config` here. So, using only the `model` can initialize the processor.

File: multimodal/tests/unittests/others/test_semantic_segmentation.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    train_img_files = get_file_paths(os.path.join(data_dir, "train/ISIC-2017_Train"))
+    train_gt_files = get_file_paths(os.path.join(data_dir, "train/ISIC-2017_Training_Part1_GroundTruth"))
+    val_img_files = get_file_paths(os.path.join(data_dir, "val/ISIC-2017_Val"))
+    val_gt_files = get_file_paths(os.path.join(data_dir, "val/ISIC-2017_Validation_Part1_GroundTruth"))
+    test_img_files = get_file_paths(os.path.join(data_dir, "test/ISIC-2017_Test"))
+    test_gt_files = get_file_paths(os.path.join(data_dir, "test/ISIC-2017_Test_v2_Part1_GroundTruth"))
+
+    train_df = pd.DataFrame({"image": train_img_files, "label": train_gt_files})
+    val_df = pd.DataFrame({"image": val_img_files, "label": val_gt_files})
+    test_df = pd.DataFrame({"image": test_img_files, "label": test_gt_files})
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/others/test_semantic_segmentation.py:44]: This block looks like duplicated in the two tests. Consider wrapping it into a util function.

File: multimodal/src/autogluon/multimodal/models/sam.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+        pred_masks = rets.pred_masks[:, 0, :, :, :]
+        pred_masks = F.interpolate(
+            pred_masks, (self.image_size, self.image_size), mode="bilinear", align_corners=False
+        )
+        if self.training:
+            return {self.prefix: {LOGITS: pred_masks}}
+        else:
+            return {self.prefix: {LOGITS: pred_masks, LABEL: batch[self.label_key]}}
+
+    def _parse_losses(self, losses):
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/sam.py:155]: When do we need `_parse_losses`?

File: multimodal/src/autogluon/multimodal/models/sam.py [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+
+        torch.save({"state_dict": self.model.state_dict()}, save_path)
+
+    def _save_configs(self, save_path=None):
+        if not save_path:
+            save_path = f"./{self.checkpoint_name}_autogluon.py"
+
+        self.config.dump(save_path)
+
+    def _load_config(self):
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/sam.py:107]: Looks like `_load_config` is not used. Remove?

File: multimodal/src/autogluon/multimodal/models/sam.py [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+        if not save_path:
+            save_path = f"./{self.checkpoint_name}_autogluon.pth"
+
+        torch.save({"state_dict": self.model.state_dict()}, save_path)
+
+    def _save_configs(self, save_path=None):
+        if not save_path:
+            save_path = f"./{self.checkpoint_name}_autogluon.py"
+
+        self.config.dump(save_path)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/sam.py:105]: Consider using huggingface's API to save the model. Looks like no need to save config? Maybe remove `_save_weights` and `_save_configs` and only keep `save`.

File: multimodal/src/autogluon/multimodal/models/sam.py [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+        self.config = self.model.config
+
+    def _load_checkpoint(self, checkpoint_name):
+        if self.pretrained:
+            self.model = SamModel.from_pretrained(checkpoint_name)
+        else:
+            configuration = SamConfig(name_or_path=checkpoint_name)
+            self.model = SamModel(configuration)
+
+    def set_data_preprocessor_device(self):
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/sam.py:80]: Where do we need `set_data_preprocessor_device`? Lightning automatically handles the devices in training.

File: multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml [hunk 17]
```diff
@@ -238,3 +238,24 @@ model:
...
+    checkpoint_name: "facebook/sam-vit-huge"
+    config_file: ""
+    image_norm: "imagenet"
+    data_types:
+      - "semantic_segmentation_img"
+    aug:
+      - "random_horizontal_flip"
+    image:
+      train_transforms:
+        - "resize_to_square"
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml:252]: `train_transforms` is for data augmentation in training.

**Review by zhiqiangdon — APPROVED** [review]
LGTM! We can merge it after the CI passes.

File: multimodal/src/autogluon/multimodal/models/sam.py [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+        self.pretrained = pretrained
+        self.checkpoint_name = checkpoint_name
+        self.frozen_layers = frozen_layers
+
+        self.device = None
+        self.name_to_id = {}
+
+        self._load_checkpoint(checkpoint_name)
+
+        freeze_model_layers(self.model, self.frozen_layers)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/sam.py:59]: What's the relation of frozen layers and peft? Will we still insert lora layers in the frozen layers?
